### PR TITLE
Kernel+Userland: Add auto-jailing symlink of dynamic loader, introduce the new `set-elf-jailed` utility

### PR DESCRIPTION
**Pull request by supercomputer7**
Relies on #22968. So this is a draft for now. The idea is to enable auto-jailing on programs by changing their dynamic loader to another which is almost identical, but enters jail-mode just before jumping to the program `main` function.

### CODE REVIEW

File: Userland/DynamicLoader/CMakeLists.txt [hunk 1]
```diff
@@ -21,3 +21,23 @@ target_link_options(Loader.so PRIVATE LINKER:--no-dynamic-linker)
 target_link_libraries(Loader.so PRIVATE NoCoverage)
 
 install(TARGETS Loader.so RUNTIME DESTINATION usr/lib/)
+
+## Build an auto-jailing version of Loader.so
+
+add_library(JailedDynamicLoader_CompileOptions INTERFACE)
```
Comment (DanShaders) [inline Userland/DynamicLoader/CMakeLists.txt:27]: I think you can just use `DynamicLoader_CompilerOptions` with `target_compile_definitions(ldjail.so PRIVATE ENTER_JAIL)` (also, no need to carry `_` prefix)

File: Userland/DynamicLoader/main.cpp [hunk 2]
```diff
@@ -149,7 +158,8 @@ NAKED void _start(int, char**, char**)
 
 ALWAYS_INLINE static void optimizer_fence()
 {
-    asm("" ::: "memory");
```
Comment (supercomputer7) [inline Userland/DynamicLoader/main.cpp:152]: Sometimes I hate `clang-format`...

File: Userland/Utilities/set-elf-jailed.cpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    if (interpreter_header.offset() != interpreter_header.vaddr().get())
+        return Error::from_errno(ENOTSUP);
+
+    auto file = TRY(Core::File::open(elf_file_path, Core::File::OpenMode::ReadWrite));
+
+    // FIXME: This is kinda hacky, but should work fine for now, as we don't
+    // really want to parse the entire ELF file and change program hadears
+    // intrusively for now.
+    auto new_path = "/usr/lib/ldjail.so\0"sv;
```
Comment (DanShaders) [inline Userland/Utilities/set-elf-jailed.cpp:55]: I think `VERIFY(interpreter_header.size_in_memory() >= new_path.size());` will be nice here.

**Review by DanShaders — COMMENTED** [review]
LGTM otherwise

File: Userland/DynamicLoader/main.cpp (hunk 2 above)
Comment (DanShaders) [reply Userland/DynamicLoader/main.cpp:152]: I'm still very confused which version of clang-format you use to get this formatting :P

**Comment by timschumi** [comment]
I see the idea, but I can't help the feeling that there are better ways to signal this to the kernel (either temporarily or permanently) than building an entirely different `DynamicLoader` and `patchelf`ing the files. Things like xattrs and environment variables come to mind.

**Comment by supercomputer7** [comment]
> I see the idea, but I can't help the feeling that there are better ways to signal this to the kernel (either temporarily or permanently) than building an entirely different `DynamicLoader` and `patchelf`ing the files. Things like xattrs and environment variables come to mind.

What about filesystems that don't support `xattr`s? Patching ELFs is a permanent solution and will work across all shells, filesystems, etc. Environment variables are also useful, but frankly it's more a temporary solution than a permanent one.

I don't mind having multiple ways to set a program into jail mode, but I think this is a completely valid way as well.

**Comment by supercomputer7** [comment]
If adding a second dynamic loader is considered a bad idea (which I don't really think it is, because it's simple and makes it easy to check if an ELF file is auto-jailed by just running `file`), we could add an option to do this in the regular dynamic loader - we will add an option in the args parsing, and an environment variable as well. We could then add a special argument in the aux vector that will signal this (secure mode come to mind) or using a special section in an ELF file for this (so `patchelf` is still needed, or some replacement of our own).

Adding support for `xattrs` is also not-a-bad idea, just not something I'd want to mess around with for now (surely not for this).

**Comment by supercomputer7** [comment]
Another good idea - make `/usr/lib/ldjail.so` a symlink to the original loader, so we don’t have 2 almost identical loaders. If you run a program through the symlink, the loader can detect this (we could add a passing of the basename of the loader in the aux vector, I think this should work fine) and then set the program into jail mode accordingly.

**Comment by supercomputer7** [comment]
> Another good idea - make `/usr/lib/ldjail.so` a symlink to the original loader, so we don’t have 2 almost identical loaders on disk. If you run a program through the symlink, the loader can detect this (we could add a passing of the basename of the loader in the aux vector, I think this should work fine) and then set the program into jail mode accordingly.

I implemented this idea and it looks really OK now.
This PR now relies on #24791 to be merged first.

**Comment by supercomputer7** [comment]
This is stale and sadly I don't have the time right now to keep it open. I might re-open it sometime in the future, if it seems like an interesting feature to someone in the project :)

**Comment by DanShaders** [comment]
Off: clearly, you haven't mastered art of removing "stale" label without making any changes for months (ref 24567) :^)

**Comment by timschumi** [comment]
> Off: clearly, you haven't mastered art of removing "stale" label without making any changes for months (ref 24567) :^)

Just saying, but that is the textbook definition of stale. Maybe not necessarily 100% for the Coroutine PR, since we also had off-GitHub discussions with a somewhat unclear outcome, but it definitely would be when applied to certain PRs that have been subject to repeated bumps.

Removing/invalidating stale labels should be reserved for cases where the PR isn't blocking on anything else but maintainer/reviewer attention. The permission to label issues and PRs isn't a free pass for keeping PRs in the queue when they shouldn't be.

**Pull request closed without merging**